File: apps/api/services/routing/polyline.py
"""Google Polyline Encoding Algorithm implementation.

Reference: https://developers.google.com/maps/documentation/utilities/polylinealgorithm
"""


def _encode_value(value: int) -> str:
    value = value << 1
    if value < 0:
        value = ~value
    chunks = []
    while value >= 0x20:
        chunks.append((0x20 | (value & 0x1F)) + 63)
        value >>= 5
    chunks.append(value + 63)
    return "".join(chr(c) for c in chunks)


def encode_polyline(coords: list[tuple[float, float]]) -> str:
    """Encode a list of (lat, lon) coordinates into a Google polyline string."""
    if not coords:
        return ""

    result = []
    prev_lat = 0
    prev_lon = 0

    for lat, lon in coords:
        lat_int = round(lat * 1e5)
        lon_int = round(lon * 1e5)
        dlat = lat_int - prev_lat
        dlon = lon_int - prev_lon
        result.append(_encode_value(dlat))
        result.append(_encode_value(dlon))
        prev_lat = lat_int
        prev_lon = lon_int

    return "".join(result)


def _decode_value(chars: str, index: int) -> tuple[int, int]:
    result = 0
    shift = 0
    while True:
        b = ord(chars[index]) - 63
        index += 1
        result |= (b & 0x1F) << shift
        shift += 5
        if b < 0x20:
            break
    if result & 1:
        result = ~result
    result >>= 1
    return result, index


def decode_polyline(polyline: str) -> list[tuple[float, float]]:
    """Decode a Google polyline string back into a list of (lat, lon) coordinates."""
    coords = []
    index = 0
    prev_lat = 0
    prev_lon = 0

    while index < len(polyline):
        dlat, index = _decode_value(polyline, index)
        dlon, index = _decode_value(polyline, index)
        prev_lat += dlat
        prev_lon += dlon
        coords.append((prev_lat / 1e5, prev_lon / 1e5))

    return coords

File: apps/web/src/utils/polyline.ts
interface Point {
  lat: number;
  lng: number;
}

function encodeValue(value: number): string {
  let v = value << 1;
  if (v < 0) v = ~v;
  const chunks: number[] = [];
  while (v >= 0x20) {
    chunks.push((0x20 | (v & 0x1f)) + 63);
    v >>= 5;
  }
  chunks.push(v + 63);
  return String.fromCharCode(...chunks);
}

export function encodePolyline(coords: Point[]): string {
  if (!coords.length) return '';
  const result: string[] = [];
  let prevLat = 0;
  let prevLng = 0;
  for (const { lat, lng } of coords) {
    const latInt = Math.round(lat * 1e5);
    const lngInt = Math.round(lng * 1e5);
    result.push(encodeValue(latInt - prevLat));
    result.push(encodeValue(lngInt - prevLng));
    prevLat = latInt;
    prevLng = lngInt;
  }
  return result.join('');
}

function decodeValue(chars: string, idx: number): [number, number] {
  let result = 0;
  let shift = 0;
  while (true) {
    const b = chars.charCodeAt(idx) - 63;
    idx += 1;
    result |= (b & 0x1f) << shift;
    shift += 5;
    if (b < 0x20) break;
  }
  if (result & 1) result = ~result;
  result >>= 1;
  return [result, idx];
}

export function decodePolyline(polyline: string): Point[] {
  const coords: Point[] = [];
  let idx = 0;
  let prevLat = 0;
  let prevLng = 0;
  while (idx < polyline.length) {
    const [dlat, next1] = decodeValue(polyline, idx);
    const [dlng, next2] = decodeValue(polyline, next1);
    prevLat += dlat;
    prevLng += dlng;
    coords.push({ lat: prevLat / 1e5, lng: prevLng / 1e5 });
    idx = next2;
  }
  return coords;
}
